### PR TITLE
baseline: assert throws Error instead of trap 

### PR DESCRIPTION
**Pull request by hzahradnik**
I changed assert so it will throw an Error. Addiotionally, i added abstract Throwable class to stdlib, which is the base class for the existing Exception and the newly introduced Error. This behavior is similar to the behavior of Java.

### CODE REVIEW

File: src/baseline/expr.rs [hunk 1]
```diff
@@ -1282,13 +1286,26 @@ where
...
+        match csite.args.get(1).unwrap() {
+            Arg::Stack(offset, _, _) => {
+                self.emit_lit_str_value(&"assert failed".to_string(), REG_RESULT);
+                self.asm
+                    .store_mem(MachineMode::Ptr, Mem::Local(*offset), REG_RESULT.into());
+            }
+            _ => panic!("unexpected argument for assert"),
+        };
+        self.emit_call_site(&csite, e.pos, REG_RESULT.into());
+        self.asm.throw(REG_RESULT, e.pos);
```
Comment (dinfuehr) [inline src/baseline/expr.rs]: At least in the new baseline compiler we should emit this code out-of-line. However you could already implement this as slow path in `baseline/asm.rs`. Both implementations should be able to share the same code so it's not additional work.

File: src/exception.rs [hunk 2]
```diff
@@ -329,25 +329,46 @@ pub fn alloc_exception(vm: &VM, msg: Ref<Str>) -> Ref<Exception> {
...
+fn set_exception_backtrace(vm: &VM, obj: Ref<Throwable>, via_retrieve: bool) {
     let stacktrace = stacktrace_from_last_dtn(vm);
     let mut obj = root(obj);
 
-    let skip = if via_retrieve { 2 } else { 0 };
+    let mut skip = 0;
+
+    // ignore every element until first not inside susubclass of Throwable (ctor of Exception)
+    if via_retrieve {
+        for elem in stacktrace.elems.iter() {
```
[the remark this answers is not in the document]
Comment (hzahradnik) [reply src/exception.rs]: I thought about that too, that with my approach we wouldn't be able to get the correct stacktrace, if the error occurs inside of a Throwable. Do you have a idea, how we could mark our magic frames/methods. Would it be sufficient to only skip constructors of Throwable subclasses?
Alternatively, would it work, if we don't call retrieve_stacktrace in the construcor, but when we actually throw the object? Then there shouldn't be a stacktrace of constructors.

[the remark this answers is not in the document]
Comment (dinfuehr) [reply src/exception.rs]: I would only skip the constructor of Throwable/Error/Exception or so. And certainly break out of the loop as soon as you find any other stack frame. Although that might require to check for the native thunk as well...

File: tests/stdlib/call-assert.dora [hunk 3]
```diff
@@ -1,5 +1,5 @@
 //= error assert
-//= output "assert failed\n0: foo(): 9\n1: call(String): 20\n2: main(): 5\n"
+//= output "uncaught exception\n0: foo(): 9\n1: call(String): 20\n2: main(): 5\n"
```
[the remark this answers is not in the document]
Comment (hzahradnik) [reply tests/stdlib/call-assert.dora]: I looked into this and especially how Java handles this. In contrast to Dora, Java doesn't handle the uncaught exception in a trap, but instead has a uncaughtException-Function on every ThreadGroup, which prints `uncaught exception in thread "main" java.lang.AssertionError`. I don't think we have to copy Java, but to print more information, we would a similar mechanism, which allows us to access the exception object. Then we could print its name and its correct stacktrace.

If it is not very important to you, I would suggest to open a separate PR.

File: src/baseline/expr.rs (hunk 1 above)
Comment (hzahradnik) [reply src/baseline/expr.rs]: Maybe I'm thinking to complicated, but if I implement it as slow path, I would to add/move a lot of functions, which I call for creating the object to asm, e.g. the `emit_lit_str_value` and `emit_call_site`. For now I will leave it as it is, but revisit it, when I'm working on the new baseline compiler.

**Pull request merged**